File: backend/auth-service/src/main/java/com/taghazout/authservice/application/dto/AuthResponse.java
package com.taghazout.authservice.application.dto;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * DTO for authentication response.
 * 
 * SOLID Principles:
 * - SRP: Only responsible for auth response data
 * - Immutable: Thread-safe
 * 
 * Returned after:
 * - Successful registration (201 Created)
 * - Successful login (200 OK)
 * - Successful token refresh (200 OK)
 * 
 * Contains:
 * - User information (email, name)
 * - JWT tokens (access + refresh)
 * 
 * JSON Example:
 * {
 * "email": "user@example.com",
 * "firstName": "John",
 * "lastName": "Doe",
 * "accessToken": "eyJ...",
 * "refreshToken": "eyJ..."
 * }
 */
public record AuthResponse(

        String email,

        @JsonProperty("firstName") String firstName,

        @JsonProperty("lastName") String lastName,

        @JsonProperty("accessToken") String accessToken,

        @JsonProperty("refreshToken") String refreshToken) {
    /**
     * Compact constructor for validation.
     */
    public AuthResponse {
        if (email == null || email.isBlank()) {
            throw new IllegalArgumentException("Email cannot be null or blank");
        }
        if (accessToken == null || accessToken.isBlank()) {
            throw new IllegalArgumentException("Access token cannot be null or blank");
        }
        if (refreshToken == null || refreshToken.isBlank()) {
            throw new IllegalArgumentException("Refresh token cannot be null or blank");
        }
    }

    /**
     * Factory method for creating response from tokens and user data.
     */
    public static AuthResponse of(String email, String firstName, String lastName,
            String accessToken, String refreshToken) {
        return new AuthResponse(email, firstName, lastName, accessToken, refreshToken);
    }

    /**
     * Factory method for creating response from tokens and email only.
     */
    public static AuthResponse of(String email, String accessToken, String refreshToken) {
        return new AuthResponse(email, null, null, accessToken, refreshToken);
    }
}

File: frontend/lib/models/auth_response.dart
/// Model class matching the backend AuthResponse DTO.
/// 
/// Returned after successful login or registration.
class AuthResponse {
  final String email;
  final String? firstName;
  final String? lastName;
  final String accessToken;
  final String? refreshToken;
  final String role; // New field
  final int? userId;

  AuthResponse({
    required this.email,
    this.firstName,
    this.lastName,
    required this.accessToken,
    this.refreshToken,
    required this.role,
    this.userId,
  });

  /// Factory constructor to create AuthResponse from JSON.
  factory AuthResponse.fromJson(Map<String, dynamic> json) {
    return AuthResponse(
      email: json['email'] as String,
      firstName: json['firstName'] as String?,
      lastName: json['lastName'] as String?,
      accessToken: json['accessToken'] as String,
      refreshToken: json['refreshToken'] as String?,
      role: json['role'] as String? ?? 'CLIENT', // Default to CLIENT if missing (backward compatibility)
      userId: json['userId'] as int?,
    );
  }

  /// Convert to JSON map.
  Map<String, dynamic> toJson() {
    return {
      'email': email,
      'firstName': firstName,
      'lastName': lastName,
      'accessToken': accessToken,
      'refreshToken': refreshToken,
      'role': role,
    };
  }
}
